### PR TITLE
Fail gracefully when installing a spec where the API is missing deps

### PR DESCRIPTION
**Pull request by segiddins**
Fixes https://github.com/bundler/bundler/issues/5426
Closes https://github.com/bundler/bundler/issues/5339

### CODE REVIEW

**Comment by indirect** [comment]
@segiddins can you explain how this fix is related to the problem in the linked ticket?

**Comment by segiddins** [comment]
Sure -- instead of deadlocking in the parallel installer, we print out that the API is returning messed up data (because the __swap__ happens after downloading the full spec) and error out suggesting using `--full-index`.

**Comment by indirect** [comment]
cool, sounds good to me. maybe we could add some debug-only output to let us figure out what API responses are bad, so we can fix them?

**Comment by segiddins** [comment]
Sure! I'll add that tomorrow -- debug or verbose output?

**Comment by segiddins** [comment]
@indirect done

File: lib/bundler/installer/gem_installer.rb [hunk 1]
```diff
@@ -16,7 +16,7 @@ def install_from_spec
       Bundler.ui.debug "#{worker}:  #{spec.name} (#{spec.version}) from #{spec.loaded_from}"
       generate_executable_stubs
       return true, post_install_message
-    rescue Bundler::InstallHookError, Bundler::SecurityError
+    rescue Bundler::InstallHookError, Bundler::SecurityError, APIResponseMismatchError
```
Comment (olleolleolle) [inline lib/bundler/installer/gem_installer.rb]: Can they all be in the "short form", without the namespace wrapped around them?

```ruby
rescue InstallHookError, SecurityError, APIResponseMismatchError
```

Comment (segiddins) [inline lib/bundler/installer/gem_installer.rb]: they could, just didn't feel like changing the other ones since they were already there

**Comment by segiddins** [comment]
Doesn't actually fix the linked-to issue :(

**Comment by segiddins** [comment]
OK, this should fix both the original issue and the first case I fixed, which is when the API has the wrong dependencies. @bradly-swart want to give this a go?

**Comment by bradly-swart** [comment]
I'll give it a whirl, thank you.

**Comment by indirect** [comment]
@segiddins I'm a little bit worried about the possible performance implications of doing this many checks on every single swap action—do you have any idea what the performance implications of this change are?

File: lib/bundler/endpoint_specification.rb [hunk 2]
```diff
@@ -91,6 +91,13 @@ def _local_specification
     end
 
     def __swap__(spec)
+      without_type = proc {|d| Gem::Dependency.new(d.name, d.requirements_list.sort) }
+      if (extra_deps = spec.runtime_dependencies.map(&without_type).-(dependencies.map(&without_type))) && extra_deps.any?
```
Comment (ivanovaleksey) [inline lib/bundler/endpoint_specification.rb]: Could you please explain to me what does `.-()` mean? 
Is it array difference [method](https://ruby-doc.org/core-2.3.3/Array.html#method-i-2D)?
Are there some pitfalls of plain `[] - []`?

Comment (agis-) [inline lib/bundler/endpoint_specification.rb]: @ivanovaleksey It's the same thing. `[] - []` is really `[].-([])`.

Comment (ivanovaleksey) [inline lib/bundler/endpoint_specification.rb]: Yes, I see. Is just matter of style? No any special caveats?

**Comment by segiddins** [comment]
> do you have any idea what the performance implications of this change are?

Given `#__swap__` is only ever caller in `Source::Rubygems#install` (and only then if the spec comes from a remote source), I don't think a map and array difference will have a noticeable performance impact (especially compared to the surrounding network & disk IO)

**Comment by indirect** [comment]
@segiddins sold. let's do it. 👍

@bundlerbot r+

**Comment by segiddins** [comment]
@bundlerbot retry

**Comment by segiddins** [comment]
@bundlerbot retry

**Pull request merged**